definitions/scenarios/self_upgrade.rb

```diff
@@ -58,15 +66,44 @@ class SelfUpgrade < SelfUpgradeBase
       manual_detection
     end
 
-    def compose
+    def downstream_self_upgrade(pkgs_to_update)
       if check_min_version('foreman', '2.5') || check_min_version('foreman-proxy', '2.5')
-        pkgs_to_update = %w[satellite-maintain rubygem-foreman_maintain]
         yum_options = req_repos_to_update_pkgs.map do |id|
           "--enablerepo=#{id}"
         end
         add_step(Procedures::Packages::Update.new(packages: pkgs_to_update, assumeyes: true,
                                                   yum_options: yum_options))
       end
     end
+
+    def upstream_self_upgrade(pkgs_to_update)
+      # This method is responsible for
+      # 1. Setup the repositories of next major version
+      # 2. Update the foreman-maintain packages from next major version repository
+      # 3. Rollback the repository to current major version
+
+      add_step(Procedures::Repositories::Setup.new(:version => upstream_target_version))
+      add_step(Procedures::Packages::Update.new(packages: pkgs_to_update, assumeyes: true))
+    ensure
```

Comment:
I'm debating if this should be `rescue` instead of `ensure`. Now it happens also after success, but that feels redundant to me.

Reply:
We need to revert the repositories to installed major version in successful run because doing the self-upgrade does not mean users will do the Foreman or Katello major upgrade just after the self-upgrade?